### ISSUE TITLE
Relocations in RO segment

### ISSUE THREAD

**Issue by aghosn** [intel/linux-sgx#430]
Hey everyone!

**TLDR** some R_X86_64_RELATIVE relocation results in a memory protection fault during enclave initialization.

I have an enclave binary with some R_X86_64_RELATIVE relocations that belong to a R/E segment. The loader `mprotect`s the corresponding addresses as writable in `CLoader::build_mem_region` as expected, but then removes write accesses too soon by calling `CLoader::set_memory_protection(false)` before the enclave initialization, and hence before the `do_relocs` function has a chance to execute, resulting in a memory error. 

It looks like the `mparser::set_memory_protection(bool)` does not take into account these pages when it loops over segments.

Thanks in advance, 
Adrien

**Comment by ZongminGu** [comment]
The current page permission changing flow aimed to handle the RELRO segment.
Do you have a simple case to help me reproduce this issue (i.e. the relocation happens inside a LOAD segment)?

BTW: Can you please tell me what is the HW you are using?

**Comment by aghosn** [comment]
Thanks for the reply! 
I do not have a simple case that I can share, but I will try to come up with one.
I have a gist [here](https://gist.github.com/aghosn/d6428c96e974fd3ca29bb52b337e621a) showing we end up with relocations inside a R/E LOAD segment.

Hardware: Intel(R) Xeon(R) CPU E3-1240 v5

**Comment by lzha101** [comment]
Thanks for reporting this issue. We will fix it in next release. Please stay tuned.
